### DISCUSSION TITLE
Official Docker Image

### DISCUSSION THREAD

**Discussion by vasily-kartashov** [jrincayc/ucblogo-code#108] [Ideas]
My plan was to go through this book: https://people.eecs.berkeley.edu/~bh/v1-toc2.html

I expected that something like `docker run -it ucblogo:latest` would give me an environment enough to check all the examples from the book. After finding out that there's no official image, I thought "how hard can it be", so that started this https://github.com/jrincayc/ucblogo-code/issues/107 :)

I am not particularly sure about the turtle graphics. If there's an option to use just the terminal for it that would be great.

On an unrelated note: I hoped to use `debian` image only to build the program, and then to run the actual program on something considerably more lightweight, like https://github.com/GoogleContainerTools/distroless/blob/main/base/README.md

**Comment by dmalec** [comment]
Ah nice! That project sounds like it lines up well with using only the terminal. Hopefully the `Dockerfile` in our discussion of #107 gets the image to the point where this works (Also, I can empathize with that moment of "how hard can it be" :) ).

One point here, the turtle commands (E.G. `FORWARD 100`) will likely cause a crash (based on my experience with running _ucblogo_ in headless environments). As long as you're not using them, shouldn't be a problem; but, just a heads up.

Given your project, another thing which may be worth mentioning is setting up an editor. There are some notes here:
https://github.com/jrincayc/ucblogo-code/wiki/using_external_editors

I just tested out with the image and was able to do the following:

1. Add `vim` to the list of packages to `apt-get install`
2. Once I'm in Logo, run the following `SETEDITOR "vim`

At that point, I can run `EDIT "foo` and `EDALL` as expected (the general pattern should work for whatever your preferred editor is). 

It would be cool to be able to run _ucblogo_ on a very lightweight container; I have to admit, I don't have much experience on that front :) One of the ideas I've been mulling (but haven't done anything on) is building a _ucblogo_ image which can take a program and capture the turtle graphics output.

**Comment by vasily-kartashov** [comment]
So this one builds an image under 10Mb.

```Dockerfile
FROM debian:10 AS builder

RUN apt-get update \
 && apt-get install -y \
    build-essential autoconf autoconf-archive automake coreutils libtool gettext git ncurses-dev texinfo texlive libwxgtk3.0-dev

WORKDIR /logo

RUN git clone https://github.com/jrincayc/ucblogo-code .
RUN autoreconf --install \
    && ./configure --without-wx-config \
    && make \
    && make install

RUN ldd ucblogo | tr -s '[:blank:]' '\n' | grep '^/' | \
    xargs -I % sh -c 'mkdir -p $(dirname deps%); cp % deps%;'


FROM gcr.io/distroless/static

COPY --from=builder /logo/deps /
COPY --from=builder /usr/local/bin/ucblogo /
COPY --from=builder /usr/local/share/ucblogo/logolib /

CMD [ "/ucblogo" ]
```

There's no vi, and "FORWARD 100" triggers a malloc, but it's a start.

**Reply by dmalec** [reply]
Nice!

I think the path to `logolib` may need to be tweaked a bit. When running the container, _ucblogo_ isn't able to find the commands defined there. The following addresses that; but, may not be the best long term approach (may be worth revisiting the `./configure` params).

```Dockerfile
COPY --from=builder /usr/local/share/ucblogo/logolib /usr/local/share/ucblogo/logolib
```

Also, I'd suggest grabbing the help file for the `HELP` command (E.G. `HELP "print`) and CSLS files for the `CSLSLOAD` command (E.G. `CSLSLOAD "algs`) :

```Dockerfile
COPY --from=builder /usr/local/share/ucblogo/helpfiles /usr/local/share/ucblogo/helpfiles
COPY --from=builder /usr/local/share/ucblogo/csls /usr/local/share/ucblogo/csls
```
